third_party/move/move-vm/transactional-tests/tests/function_values_safety/closure_assign_return.masm

```diff
@@ -0,0 +1,55 @@
+//# publish
+module 0x66::helpers
+
+// Uses pseudo captured structure to force a constructed
+// closure to have abilities. Note that declaration of
+// types (like return type) do not matter for paranoid
+// mode, rather it uses the actual type of the values
+// on the stack.
+
+public fun f_copy_drop_store(): |u64|u32 has copy+drop+store
+    pack_closure action, 0
+    ret
+
+public fun f(): |u64|u32 has drop
+    pack_closure action, 0
+    ret
+
+#[persistent] fun action(x: u64): u32
+    move_loc x
+    cast_u32
+    ld_u32 20
+    add
+    ret
+
+public fun assert_eq(x: u32, y:u32)
+    move_loc x
+    move_loc y
+    eq
+    br_true r
+    ld_u64 255
+    abort
+ r: ret
+
+
+//# publish
+module 0x66::ok
+use 0x66::helpers
+
+fun assignable(): (|u64|u32, |u64|u32 has copy+drop)
+    call helpers::f
+    call helpers::f_copy_drop_store
+    ret
+
+//# run --verbose 0x66::ok::assignable
+
+//# publish
+module 0x66::not_ok
+use 0x66::helpers
+
+fun assignable(): (|u64|u32, |u64|u32 has copy+drop)
+    call helpers::f_copy_drop_store
+    call helpers::f
```

Comment:
So without your change this is not detected? I think the main issue is that `assignable` here is an entry point, and we do not type final return:
```
// in interpreter.rs for ExitCode::Return
} else {
    return Ok(self.operand_stack.value);
}
```
so seems like we just need to type check there, not? Otherwise if it is not the last frame next user type checks assignability.

Reply:
No, we never type check return, whether entry point or not. We only check if the returned value is stored in a local or passed to some other functions. But the function can declare a very different return type then we check, and we simply ignore this type.